### PR TITLE
test: FairOS

### LINKED ISSUE #108: Integration test for FairOS
The main operations:
- pod listing
- directory listing
- authentification

Have to be covered by tests.

### PR DESCRIPTION
**Pull request by IgorShadurin**
Tested described in #108 are implemented.

Our sharing feature implementation differs from FairOS https://github.com/fairDataSociety/fdp-storage/issues/152. So I could cover it after implementation fix.

Close #108

### CODE REVIEW

File: test/integration/fdp-class.fairos.spec.ts [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+        name_hash: nameHash,
+        public_key: publicKey,
+        message: 'user signed-up successfully',
+      })
+
+      const data = await fdp.account.login(user.username, user.password)
+      expect(data.address).toEqual(user.address)
+    })
+
+    it('import account to fairos', async () => {
```
Comment (nugaon) [inline test/integration/fdp-class.fairos.spec.ts:68]: this test is a bit vague for me. it just calls the register API EP on FairOS for a user but nothing else.

File: test/utils.ts [hunk 2]
```diff
@@ -181,3 +182,61 @@ export function getCachedBatchId(): BatchId {
...
+      // eslint-disable-next-line no-empty
+    } catch (e) {}
+
+    if (text === 'OK\n') {
+      return
+    } else if (text) {
+      throw new Error('Incorrect FairOS API answer')
+    }
+
+    await sleep(3000)
```
Comment (nugaon) [inline test/utils.ts:211]: it is a bit long waiting time, maybe 1sec?

File: test/utils.ts [hunk 3]
```diff
@@ -181,3 +182,61 @@ export function getCachedBatchId(): BatchId {
...
+  for (let i = 0; i <= 100; i++) {
+    let text
+    try {
+      text = (await axios.get(url)).data
+      // eslint-disable-next-line no-empty
+    } catch (e) {}
+
+    if (text === 'OK\n') {
+      return
+    } else if (text) {
```
Comment (nugaon) [inline test/utils.ts:207]: I get back 
```
FairOS-dfs
0.7.3-dev
/root/.fairOS/dfs
http://fdp-play-queen:1633
http://fdp-play-queen:1635
trace
:9090
:9091
localhost
[*]
```
from fairos for this request, that still does not seem wrong as a response, but this logic just tests against `OK\n`.
maybe you can check against response status instead?

File: test/integration/fdp-class.fairos.spec.ts [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+      const podName1 = generateRandomHexString()
+      const podName2 = generateRandomHexString()
+      const podName3 = generateRandomHexString()
+
+      await fairos.registerV1(user.username, user.password, user.mnemonic)
+      const createResponse = await fairos.podNew(podName1, user.password)
+      expect(createResponse.status).toEqual(201)
+      expect(createResponse.data).toStrictEqual({ message: 'pod created successfully' })
+
+      await fdp.account.setAccountFromMnemonic(user.mnemonic)
```
Comment (nugaon) [inline test/integration/fdp-class.fairos.spec.ts:117]: unnecessary await.

File: test/integration/fdp-class.fairos.spec.ts [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+      const user = generateUser(fdp)
+      const podName1 = generateRandomHexString()
+      const directoryName1 = generateRandomHexString()
+      const fullDirectoryName1 = '/' + directoryName1
+      const subDirectoryName1 = generateRandomHexString()
+      const fullSubDirectoryName1 = fullDirectoryName1 + '/' + subDirectoryName1
+      const directoryName2 = generateRandomHexString()
+      const fullDirectoryName2 = '/' + directoryName2
+
+      await fdp.account.setAccountFromMnemonic(user.mnemonic)
```
Comment (nugaon) [inline test/integration/fdp-class.fairos.spec.ts:223]: unnecessary await

**Review by nugaon — CHANGES_REQUESTED** [review]
the test cases seem right, but I got different response for FairOS API check than you, see my comment below.

File: test/utils.ts (hunk 3 above)
Comment (IgorShadurin) [reply test/utils.ts:207]: hm, on my side after `fdp-play start --pull --fresh --fairos` I receive `0.8.2-rc8-609cfb8`

But yes, checking the status is a more general approach for different versions

**Review by nugaon — COMMENTED** [review]
I checked now and I get 4 errors on FairOS tests:
● Pod › should create pods in fairos and list them in fdp
and
● Pod › should delete pod in fairos and it will disappear in fdp
throw error because the response message has additional `code: 201`. IMO this could be taken out from the response on FairOS since it should be set in the response status instead.

About the other two erros I'm more anxious:
  ● Account › should register in fdp and login in 
and 
  ● Account › should register in fairos and login in fdp

both have error message
    `Request failed with status code 404`

File: test/integration/fdp-class.fairos.spec.ts [hunk 6]
```diff
@@ -61,35 +59,23 @@ describe('Fair Data Protocol with FairOS-dfs', () => {
...
-        index: 1,
-        name: podName1,
-        password: bytesToHex(createdPod1.password),
-      }
       expect(response.status).toEqual(200)
       expect(response.data).toStrictEqual({
-        pods: [expectedPod1],
-        sharedPods: [],
+        podName: [podName1],
+        sharedPodName: [],
```
Comment (nugaon) [inline test/integration/fdp-class.fairos.spec.ts:68]: when did it become `podName` and `sharedPodName` from `pods` and `sharedPodName`?
https://github.com/fairDataSociety/FIPs/pull/61/files#diff-81f0c568292734f60520b0a790db67b5c50b148490ed1fd7ccd95283172be861R86-R89

Comment (asabya) [reply test/integration/fdp-class.fairos.spec.ts:68]: will change them back.

**Review by nugaon — APPROVED** [review]
LGTM!

**Pull request merged**